// @bazel/typescript/tsetse/checker.js
/**
 * @fileoverview Checker contains all the information we need to perform source
 * file AST traversals and report errors.
 */
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "typescript", "./failure"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ts = require("typescript");
    var failure_1 = require("./failure");
    /**
     * Tsetse rules use on() and addFailureAtNode() for rule implementations.
     * Rules can get a ts.TypeChecker from checker.typeChecker so typed rules are
     * possible. Compiler uses execute() to run the Tsetse check.
     */
    var Checker = /** @class */ (function () {
        function Checker(program) {
            /**
             * nodeHandlersMap contains node to handlers mapping for all enabled rules.
             */
            this.nodeHandlersMap = new Map();
            this.failures = [];
            // currentCode will be set before invoking any handler functions so the value
            // initialized here is never used.
            this.currentCode = 0;
            // Avoid the cost for each rule to create a new TypeChecker.
            this.typeChecker = program.getTypeChecker();
        }
        /**
         * This doesn't run any checks yet. Instead, it registers `handlerFunction` on
         * `nodeKind` node in `nodeHandlersMap` map. After all rules register their
         * handlers, the source file AST will be traversed.
         */
        Checker.prototype.on = function (nodeKind, handlerFunction, code) {
            var newHandler = { handlerFunction: handlerFunction, code: code };
            var registeredHandlers = this.nodeHandlersMap.get(nodeKind);
            if (registeredHandlers === undefined) {
                this.nodeHandlersMap.set(nodeKind, [newHandler]);
            }
            else {
                registeredHandlers.push(newHandler);
            }
        };
        /**
         * Add a failure with a span. addFailure() is currently private because
         * `addFailureAtNode` is preferred.
         */
        Checker.prototype.addFailure = function (start, end, failureText) {
            if (!this.currentSourceFile) {
                throw new Error('Source file not defined');
            }
            if (start >= end || end > this.currentSourceFile.end || start < 0) {
                // Since only addFailureAtNode() is exposed for now this shouldn't happen.
                throw new Error("Invalid start and end position: [" + start + ", " + end + "]" +
                    (" in file " + this.currentSourceFile.fileName + "."));
            }
            var failure = new failure_1.Failure(this.currentSourceFile, start, end, failureText, this.currentCode);
            this.failures.push(failure);
        };
        Checker.prototype.addFailureAtNode = function (node, failureText) {
            // node.getStart() takes a sourceFile as argument whereas node.getEnd()
            // doesn't need it.
            this.addFailure(node.getStart(this.currentSourceFile), node.getEnd(), failureText);
        };
        /**
         * Walk `sourceFile`, invoking registered handlers with Checker as the first
         * argument and current node as the second argument. Return failures if there
         * are any.
         */
        Checker.prototype.execute = function (sourceFile) {
            var thisChecker = this;
            this.currentSourceFile = sourceFile;
            this.failures = [];
            ts.forEachChild(sourceFile, run);
            return this.failures;
            function run(node) {
                var e_1, _a;
                var handlers = thisChecker.nodeHandlersMap.get(node.kind);
                if (handlers !== undefined) {
                    try {
                        for (var handlers_1 = __values(handlers), handlers_1_1 = handlers_1.next(); !handlers_1_1.done; handlers_1_1 = handlers_1.next()) {
                            var handler = handlers_1_1.value;
                            thisChecker.currentCode = handler.code;
                            handler.handlerFunction(thisChecker, node);
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (handlers_1_1 && !handlers_1_1.done && (_a = handlers_1.return)) _a.call(handlers_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                ts.forEachChild(node, run);
            }
        };
        return Checker;
    }());
    exports.Checker = Checker;
});
