// @bazel/typescript/tsetse/rules/ban_promise_as_condition_rule.js
/**
 * @fileoverview Bans using a promise as a condition. Promises are always
 * truthy, and this pattern is likely to be a bug where the developer meant
 * if(await returnsPromise()) {} and forgot the await.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tsutils", "typescript", "../error_code", "../rule"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tsutils = require("tsutils");
    var ts = require("typescript");
    var error_code_1 = require("../error_code");
    var rule_1 = require("../rule");
    var Rule = /** @class */ (function (_super) {
        __extends(Rule, _super);
        function Rule() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.ruleName = 'ban-promise-as-condition';
            _this.code = error_code_1.ErrorCode.BAN_PROMISE_AS_CONDITION;
            return _this;
        }
        Rule.prototype.register = function (checker) {
            checker.on(ts.SyntaxKind.ConditionalExpression, checkConditional, this.code);
            checker.on(ts.SyntaxKind.BinaryExpression, checkBinaryExpression, this.code);
            checker.on(ts.SyntaxKind.WhileStatement, checkWhileStatement, this.code);
            checker.on(ts.SyntaxKind.IfStatement, checkIfStatement, this.code);
        };
        return Rule;
    }(rule_1.AbstractRule));
    exports.Rule = Rule;
    /** Error message to display. */
    function thenableText(nodeType, isVariable) {
        return "Found a thenable " + (isVariable ? 'variable' : 'return value') + " being" +
            (" used as " + nodeType + ". Promises are always truthy, await the value to get") +
            ' a boolean value.';
    }
    function thenableVariableText(nodeType) {
        return thenableText(nodeType, true);
    }
    function thenableReturnText(nodeType) {
        return thenableText(nodeType, false);
    }
    /** Ternary: prom ? y : z */
    function checkConditional(checker, node) {
        addFailureIfThenableCallExpression(checker, node.condition, thenableReturnText('a conditional'));
        addFailureIfThenableIdentifier(checker, node.condition, thenableVariableText('a conditional'));
    }
    /**
     *  Binary expression: prom || y or prom && y. Only check left side because
     *  myThing && myThing.prom seems legitimate.
     */
    function checkBinaryExpression(checker, node) {
        if (node.operatorToken.kind !== ts.SyntaxKind.BarBarToken &&
            node.operatorToken.kind !== ts.SyntaxKind.AmpersandAmpersandToken) {
            return;
        }
        addFailureIfThenableCallExpression(checker, node.left, thenableReturnText('a binary expression'));
        addFailureIfThenableIdentifier(checker, node.left, thenableVariableText('a binary expression'));
    }
    /** While statement: while (prom) {} */
    function checkWhileStatement(checker, node) {
        addFailureIfThenableCallExpression(checker, node.expression, thenableReturnText('a while statement'));
        addFailureIfThenableIdentifier(checker, node.expression, thenableVariableText('a while  statement'));
    }
    /** If statement: if (prom) {} */
    function checkIfStatement(checker, node) {
        addFailureIfThenableCallExpression(checker, node.expression, thenableReturnText('an if statement'));
        addFailureIfThenableIdentifier(checker, node.expression, thenableVariableText('an if statement'));
    }
    /** Helper methods */
    function addFailureIfThenableCallExpression(checker, callExpression, errorMessage) {
        if (!tsutils.isCallExpression(callExpression)) {
            return;
        }
        var typeChecker = checker.typeChecker;
        var signature = typeChecker.getResolvedSignature(callExpression);
        // Return value of getResolvedSignature is `Signature | undefined` in ts 3.1
        // so we must check if the return value is valid to compile with ts 3.1.
        if (!signature) {
            throw new Error('Unexpected undefined signature for call expression');
        }
        var returnType = typeChecker.getReturnTypeOfSignature(signature);
        if (isNonFalsyThenableType(typeChecker, callExpression, returnType)) {
            checker.addFailureAtNode(callExpression, errorMessage);
        }
    }
    function addFailureIfThenableIdentifier(checker, identifier, errorMessage) {
        if (!tsutils.isIdentifier(identifier)) {
            return;
        }
        if (isNonFalsyThenableType(checker.typeChecker, identifier)) {
            checker.addFailureAtNode(identifier, errorMessage);
        }
    }
    /**
     * If the type is a union type and has a falsy part it may be legitimate to use
     * it as a condition, so allow those through. (e.g. Promise<boolean> | boolean)
     * Otherwise, check if it's thenable. If so it should be awaited.
     */
    function isNonFalsyThenableType(typeChecker, node, type) {
        if (type === void 0) { type = typeChecker.getTypeAtLocation(node); }
        if (hasFalsyParts(typeChecker.getTypeAtLocation(node))) {
            return false;
        }
        return tsutils.isThenableType(typeChecker, node, type);
    }
    function hasFalsyParts(type) {
        var typeParts = tsutils.unionTypeParts(type);
        var hasFalsyParts = typeParts.filter(function (part) { return tsutils.isFalsyType(part); }).length > 0;
        return hasFalsyParts;
    }
});
