// @bazel/typescript/tsetse/rules/check_return_value_rule.js
/**
 * @fileoverview A Tsetse rule that checks the return value of certain functions
 * must be used.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tsutils", "typescript", "../error_code", "../rule"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tsutils = require("tsutils");
    var ts = require("typescript");
    var error_code_1 = require("../error_code");
    var rule_1 = require("../rule");
    var FAILURE_STRING = 'return value is unused.'
        + '\n\tSee http://tsetse.info/check-return-value';
    // A list of well-known functions that the return value must be used. If unused
    // then the function call is either a no-op (e.g. 'foo.trim()' foo is unchanged)
    // or can be replaced by another (Array.map() should be replaced with a loop or
    // Array.forEach() if the return value is unused).
    var METHODS_TO_CHECK = new Set([
        ['Array', 'concat'],
        ['Array', 'filter'],
        ['Array', 'map'],
        ['Array', 'slice'],
        ['Function', 'bind'],
        ['Object', 'create'],
        ['string', 'concat'],
        ['string', 'normalize'],
        ['string', 'padStart'],
        ['string', 'padEnd'],
        ['string', 'repeat'],
        ['string', 'slice'],
        ['string', 'split'],
        ['string', 'substr'],
        ['string', 'substring'],
        ['string', 'toLocaleLowerCase'],
        ['string', 'toLocaleUpperCase'],
        ['string', 'toLowerCase'],
        ['string', 'toUpperCase'],
        ['string', 'trim'],
    ].map(function (list) { return list.join('#'); }));
    var Rule = /** @class */ (function (_super) {
        __extends(Rule, _super);
        function Rule() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.ruleName = 'check-return-value';
            _this.code = error_code_1.ErrorCode.CHECK_RETURN_VALUE;
            return _this;
        }
        // registers checkCallExpression() function on ts.CallExpression node.
        // TypeScript conformance will traverse the AST of each source file and run
        // checkCallExpression() every time it encounters a ts.CallExpression node.
        Rule.prototype.register = function (checker) {
            checker.on(ts.SyntaxKind.CallExpression, checkCallExpression, this.code);
        };
        return Rule;
    }(rule_1.AbstractRule));
    exports.Rule = Rule;
    function checkCallExpression(checker, node) {
        // Short-circuit before using the typechecker if possible, as its expensive.
        // Workaround for https://github.com/Microsoft/TypeScript/issues/27997
        if (tsutils.isExpressionValueUsed(node)) {
            return;
        }
        // Check if this CallExpression is one of the well-known functions and returns
        // a non-void value that is unused.
        var signature = checker.typeChecker.getResolvedSignature(node);
        if (signature !== undefined) {
            var returnType = checker.typeChecker.getReturnTypeOfSignature(signature);
            if (!!(returnType.flags & ts.TypeFlags.Void)) {
                return;
            }
            // Although hasCheckReturnValueJsDoc() is faster than isBlackListed(), it
            // returns false most of the time and thus isBlackListed() would have to run
            // anyway. Therefore we short-circuit hasCheckReturnValueJsDoc().
            if (!isBlackListed(node, checker.typeChecker) &&
                !hasCheckReturnValueJsDoc(node, checker.typeChecker)) {
                return;
            }
            checker.addFailureAtNode(node, FAILURE_STRING);
        }
    }
    function isBlackListed(node, tc) {
        switch (node.expression.kind) {
            case ts.SyntaxKind.PropertyAccessExpression:
            case ts.SyntaxKind.ElementAccessExpression:
                // Example: foo.bar() or foo[bar]()
                // expressionNode is foo
                var nodeExpression = node.expression.expression;
                var nodeExpressionString = nodeExpression.getText();
                var nodeType = tc.getTypeAtLocation(nodeExpression);
                // nodeTypeString is the string representation of the type of foo
                var nodeTypeString = tc.typeToString(nodeType);
                if (nodeTypeString.endsWith('[]')) {
                    nodeTypeString = 'Array';
                }
                if (nodeTypeString === 'ObjectConstructor') {
                    nodeTypeString = 'Object';
                }
                if (tsutils.isTypeFlagSet(nodeType, ts.TypeFlags.StringLiteral)) {
                    nodeTypeString = 'string';
                }
                // nodeFunction is bar
                var nodeFunction = '';
                if (tsutils.isPropertyAccessExpression(node.expression)) {
                    nodeFunction = node.expression.name.getText();
                }
                if (tsutils.isElementAccessExpression(node.expression)) {
                    var argument = node.expression.argumentExpression;
                    if (argument !== undefined) {
                        nodeFunction = argument.getText();
                    }
                }
                // Check if 'foo#bar' or `${typeof foo}#bar` is in the blacklist.
                if (METHODS_TO_CHECK.has(nodeTypeString + "#" + nodeFunction) ||
                    METHODS_TO_CHECK.has(nodeExpressionString + "#" + nodeFunction)) {
                    return true;
                }
                // For 'str.replace(regexp|substr, newSubstr|function)' only check when
                // the second parameter is 'newSubstr'.
                if ((nodeTypeString + "#" + nodeFunction === 'string#replace') ||
                    (nodeExpressionString + "#" + nodeFunction === 'string#replace')) {
                    return node.arguments.length === 2 &&
                        !tsutils.isFunctionWithBody(node.arguments[1]);
                }
                break;
            case ts.SyntaxKind.Identifier:
                // Example: foo()
                // We currently don't have functions of this kind in blacklist.
                var identifier = node.expression;
                if (METHODS_TO_CHECK.has(identifier.text)) {
                    return true;
                }
                break;
            default:
                break;
        }
        return false;
    }
    function hasCheckReturnValueJsDoc(node, tc) {
        var e_1, _a;
        var symbol = tc.getSymbolAtLocation(node.expression);
        if (symbol === undefined) {
            return false;
        }
        if (tsutils.isSymbolFlagSet(symbol, ts.SymbolFlags.Alias)) {
            symbol = tc.getAliasedSymbol(symbol);
        }
        try {
            for (var _b = __values(symbol.getJsDocTags()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var jsDocTagInfo = _c.value;
                if (jsDocTagInfo.name === 'checkReturnValue') {
                    return true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return false;
    }
});
