// @bazel/typescript/tsc_wrapped/cache.js
/**
 * @license
 * Copyright 2017 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "fs", "typescript", "./perf_trace"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var fs = require("fs");
    var ts = require("typescript");
    var perfTrace = require("./perf_trace");
    /**
     * Cache exposes a trivial LRU cache.
     *
     * This code uses the fact that JavaScript hash maps are linked lists - after
     * reaching the cache size limit, it deletes the oldest (first) entries. Used
     * cache entries are moved to the end of the list by deleting and re-inserting.
     */
    var Cache = /** @class */ (function () {
        function Cache(name, debug) {
            this.name = name;
            this.debug = debug;
            this.map = new Map();
            this.stats = { reads: 0, hits: 0, evictions: 0 };
        }
        Cache.prototype.set = function (key, value) {
            this.map.set(key, value);
        };
        Cache.prototype.get = function (key, updateCache) {
            if (updateCache === void 0) { updateCache = true; }
            this.stats.reads++;
            var entry = this.map.get(key);
            if (updateCache) {
                if (entry) {
                    this.debug(this.name, 'cache hit:', key);
                    this.stats.hits++;
                    // Move an entry to the end of the cache by deleting and re-inserting
                    // it.
                    this.map.delete(key);
                    this.map.set(key, entry);
                }
                else {
                    this.debug(this.name, 'cache miss:', key);
                }
                this.traceStats();
            }
            return entry;
        };
        Cache.prototype.delete = function (key) {
            this.map.delete(key);
        };
        Cache.prototype.evict = function (unevictableKeys) {
            var e_1, _a;
            // Drop half the cache, the least recently used entry == the first entry.
            this.debug('Evicting from the', this.name, 'cache...');
            var originalSize = this.map.size;
            var numberKeysToDrop = originalSize / 2;
            if (numberKeysToDrop === 0) {
                return 0;
            }
            try {
                // Map keys are iterated in insertion order, since we reinsert on access
                // this is indeed a LRU strategy.
                // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/keys
                for (var _b = __values(this.map.keys()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if (numberKeysToDrop === 0)
                        break;
                    if (unevictableKeys && unevictableKeys.has(key))
                        continue;
                    this.map.delete(key);
                    numberKeysToDrop--;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var keysDropped = originalSize - this.map.size;
            this.stats.evictions += keysDropped;
            this.debug('Evicted', keysDropped, this.name, 'cache entries');
            this.traceStats();
            return keysDropped;
        };
        Cache.prototype.keys = function () {
            return this.map.keys();
        };
        Cache.prototype.resetStats = function () {
            this.stats = { hits: 0, reads: 0, evictions: 0 };
        };
        Cache.prototype.printStats = function () {
            var percentage;
            if (this.stats.reads === 0) {
                percentage = 100.00; // avoid "NaN %"
            }
            else {
                percentage = (this.stats.hits / this.stats.reads * 100).toFixed(2);
            }
            this.debug(this.name + " cache stats: " + percentage + "% hits", this.stats);
        };
        Cache.prototype.traceStats = function () {
            var _a;
            // counters are rendered as stacked bar charts, so record cache
            // hits/misses rather than the 'reads' stat tracked in stats
            // so the chart makes sense.
            perfTrace.counter(this.name + " cache hit rate", {
                'hits': this.stats.hits,
                'misses': this.stats.reads - this.stats.hits,
            });
            perfTrace.counter(this.name + " cache evictions", {
                'evictions': this.stats.evictions,
            });
            perfTrace.counter(this.name + " cache size", (_a = {},
                _a[this.name + "s"] = this.map.size,
                _a));
        };
        return Cache;
    }());
    /**
     * Default memory size, beyond which we evict from the cache.
     */
    var DEFAULT_MAX_MEM_USAGE = 1024 * (1 << 20 /* 1 MB */);
    /**
     * FileCache is a trivial LRU cache for typescript-parsed bazel-output files.
     *
     * Cache entries include an opaque bazel-supplied digest to track staleness.
     * Expected digests must be set (using updateCache) before using the cache.
     */
    // TODO(martinprobst): Drop the <T> parameter, it's no longer used.
    var FileCache = /** @class */ (function () {
        function FileCache(debug) {
            var _this = this;
            this.debug = debug;
            this.fileCache = new Cache('file', this.debug);
            /**
             * FileCache does not know how to construct bazel's opaque digests. This
             * field caches the last (or current) compile run's digests, so that code
             * below knows what digest to assign to a newly loaded file.
             */
            this.lastDigests = new Map();
            /**
             * FileCache can enter a degenerate state, where all cache entries are pinned
             * by lastDigests, but the system is still out of memory. In that case, do not
             * attempt to free memory until lastDigests has changed.
             */
            this.cannotEvict = false;
            /**
             * Because we cannot measuse the cache memory footprint directly, we evict
             * when the process' total memory usage goes beyond this number.
             */
            this.maxMemoryUsage = DEFAULT_MAX_MEM_USAGE;
            /**
             * Returns whether the cache should free some memory.
             *
             * Defined as a property so it can be overridden in tests.
             */
            this.shouldFreeMemory = function () {
                return process.memoryUsage().heapUsed > _this.maxMemoryUsage;
            };
        }
        FileCache.prototype.setMaxCacheSize = function (maxCacheSize) {
            if (maxCacheSize < 0) {
                throw new Error("FileCache max size is negative: " + maxCacheSize);
            }
            this.debug('Cache max size is', maxCacheSize >> 20, 'MB');
            this.maxMemoryUsage = maxCacheSize;
            this.maybeFreeMemory();
        };
        FileCache.prototype.resetMaxCacheSize = function () {
            this.setMaxCacheSize(DEFAULT_MAX_MEM_USAGE);
        };
        FileCache.prototype.updateCache = function (digests) {
            var e_2, _a;
            // TODO(martinprobst): drop the Object based version, it's just here for
            // backwards compatibility.
            if (!(digests instanceof Map)) {
                digests = new Map(Object.keys(digests).map(function (k) { return [k, digests[k]]; }));
            }
            this.debug('updating digests:', digests);
            this.lastDigests = digests;
            this.cannotEvict = false;
            try {
                for (var _b = __values(digests.entries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var _d = __read(_c.value, 2), filePath = _d[0], newDigest = _d[1];
                    var entry = this.fileCache.get(filePath, /*updateCache=*/ false);
                    if (entry && entry.digest !== newDigest) {
                        this.debug('dropping file cache entry for', filePath, 'digests', entry.digest, newDigest);
                        this.fileCache.delete(filePath);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        };
        FileCache.prototype.getLastDigest = function (filePath) {
            var digest = this.lastDigests.get(filePath);
            if (!digest) {
                throw new Error("missing input digest for " + filePath + "." +
                    ("(only have " + Array.from(this.lastDigests.keys()) + ")"));
            }
            return digest;
        };
        FileCache.prototype.getCache = function (filePath) {
            var entry = this.fileCache.get(filePath);
            if (entry)
                return entry.value;
            return undefined;
        };
        FileCache.prototype.putCache = function (filePath, entry) {
            var dropped = this.maybeFreeMemory();
            this.fileCache.set(filePath, entry);
            this.debug('Loaded file:', filePath, 'dropped', dropped, 'files');
        };
        /**
         * Returns true if the given filePath was reported as an input up front and
         * has a known cache digest. FileCache can only cache known files.
         */
        FileCache.prototype.isKnownInput = function (filePath) {
            return this.lastDigests.has(filePath);
        };
        FileCache.prototype.inCache = function (filePath) {
            return !!this.getCache(filePath);
        };
        FileCache.prototype.resetStats = function () {
            this.fileCache.resetStats();
        };
        FileCache.prototype.printStats = function () {
            this.fileCache.printStats();
        };
        FileCache.prototype.traceStats = function () {
            this.fileCache.traceStats();
        };
        /**
         * Frees memory if required. Returns the number of dropped entries.
         */
        FileCache.prototype.maybeFreeMemory = function () {
            if (!this.shouldFreeMemory() || this.cannotEvict) {
                return 0;
            }
            var dropped = this.fileCache.evict(this.lastDigests);
            if (dropped === 0) {
                // Freeing memory did not drop any cache entries, because all are pinned.
                // Stop evicting until the pinned list changes again. This prevents
                // degenerating into an O(n^2) situation where each file load iterates
                // through the list of all files, trying to evict cache keys in vain
                // because all are pinned.
                this.cannotEvict = true;
            }
            return dropped;
        };
        FileCache.prototype.getFileCacheKeysForTest = function () {
            return Array.from(this.fileCache.keys());
        };
        return FileCache;
    }());
    exports.FileCache = FileCache;
    /**
     * ProgramAndFileCache is a trivial LRU cache for typescript-parsed programs and
     * bazel-output files.
     *
     * Programs are evicted before source files because they have less reuse across
     * compilations.
     */
    var ProgramAndFileCache = /** @class */ (function (_super) {
        __extends(ProgramAndFileCache, _super);
        function ProgramAndFileCache() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.programCache = new Cache('program', _this.debug);
            return _this;
        }
        ProgramAndFileCache.prototype.getProgram = function (target) {
            return this.programCache.get(target);
        };
        ProgramAndFileCache.prototype.putProgram = function (target, program) {
            var dropped = this.maybeFreeMemory();
            this.programCache.set(target, program);
            this.debug('Loaded program:', target, 'dropped', dropped, 'entries');
        };
        ProgramAndFileCache.prototype.resetStats = function () {
            _super.prototype.resetStats.call(this);
            this.programCache.resetStats();
        };
        ProgramAndFileCache.prototype.printStats = function () {
            _super.prototype.printStats.call(this);
            this.programCache.printStats();
        };
        ProgramAndFileCache.prototype.traceStats = function () {
            _super.prototype.traceStats.call(this);
            this.programCache.traceStats();
        };
        ProgramAndFileCache.prototype.maybeFreeMemory = function () {
            if (!this.shouldFreeMemory())
                return 0;
            var dropped = this.programCache.evict();
            if (dropped > 0)
                return dropped;
            return _super.prototype.maybeFreeMemory.call(this);
        };
        ProgramAndFileCache.prototype.getProgramCacheKeysForTest = function () {
            return Array.from(this.programCache.keys());
        };
        return ProgramAndFileCache;
    }(FileCache));
    exports.ProgramAndFileCache = ProgramAndFileCache;
    /**
     * Load a source file from disk, or possibly return a cached version.
     */
    var CachedFileLoader = /** @class */ (function () {
        // TODO(alexeagle): remove unused param after usages updated:
        // angular:packages/bazel/src/ngc-wrapped/index.ts
        function CachedFileLoader(cache, unused) {
            this.cache = cache;
            /** Total amount of time spent loading files, for the perf trace. */
            this.totalReadTimeMs = 0;
        }
        CachedFileLoader.prototype.fileExists = function (filePath) {
            return this.cache.isKnownInput(filePath);
        };
        CachedFileLoader.prototype.loadFile = function (fileName, filePath, langVer) {
            var sourceFile = this.cache.getCache(filePath);
            if (!sourceFile) {
                var readStart = Date.now();
                var sourceText = fs.readFileSync(filePath, 'utf8');
                sourceFile = ts.createSourceFile(fileName, sourceText, langVer, true);
                var entry = {
                    digest: this.cache.getLastDigest(filePath),
                    value: sourceFile
                };
                var readEnd = Date.now();
                this.cache.putCache(filePath, entry);
                this.totalReadTimeMs += readEnd - readStart;
                perfTrace.counter('file load time', {
                    'read': this.totalReadTimeMs,
                });
                perfTrace.snapshotMemoryUsage();
            }
            return sourceFile;
        };
        return CachedFileLoader;
    }());
    exports.CachedFileLoader = CachedFileLoader;
    /** Load a source file from disk. */
    var UncachedFileLoader = /** @class */ (function () {
        function UncachedFileLoader() {
        }
        UncachedFileLoader.prototype.fileExists = function (filePath) {
            return ts.sys.fileExists(filePath);
        };
        UncachedFileLoader.prototype.loadFile = function (fileName, filePath, langVer) {
            var sourceText = fs.readFileSync(filePath, 'utf8');
            return ts.createSourceFile(fileName, sourceText, langVer, true);
        };
        return UncachedFileLoader;
    }());
    exports.UncachedFileLoader = UncachedFileLoader;
});
