// @bazel/typescript/tsc_wrapped/diagnostics.js
/**
 * @fileoverview extensions to TypeScript functionality around error handling
 * (ts.Diagnostics).
 */
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ts = require("typescript");
    /**
     * If the current compilation was a compilation test expecting certain
     * diagnostics, filter out the expected diagnostics, and add new diagnostics
     * (aka errors) for non-matched diagnostics.
     */
    function filterExpected(bazelOpts, diagnostics, formatFn) {
        if (formatFn === void 0) { formatFn = uglyFormat; }
        if (!bazelOpts.expectedDiagnostics.length)
            return diagnostics;
        // The regex contains two parts:
        // 1. Optional position: '\(5,1\)'
        // 2. Required TS error: 'TS2000: message text.'
        // Need triple escapes because the expected diagnostics that we're matching
        // here are regexes, too.
        var ERROR_RE = /^(?:\\\((\d*),(\d*)\\\).*)?TS(\d+):(.*)/;
        var incorrectErrors = bazelOpts.expectedDiagnostics.filter(function (e) { return !e.match(ERROR_RE); });
        if (incorrectErrors.length) {
            var msg = "Expected errors must match regex " + ERROR_RE + "\n\t" +
                ("expected errors are \"" + incorrectErrors.join(', ') + "\"");
            return [{
                    file: undefined,
                    start: 0,
                    length: 0,
                    messageText: msg,
                    category: ts.DiagnosticCategory.Error,
                    code: 0,
                }];
        }
        var expectedDiags = bazelOpts.expectedDiagnostics.map(function (expected) {
            var m = expected.match(/^(?:\\\((\d*),(\d*)\\\).*)?TS(\d+):(.*)$/);
            if (!m) {
                throw new Error('Incorrect expected error, did you forget character escapes in ' +
                    expected);
            }
            var _a = __read(m, 5), lineStr = _a[1], columnStr = _a[2], codeStr = _a[3], regexp = _a[4];
            var _b = __read([lineStr, columnStr, codeStr].map(function (str) {
                var i = Number(str);
                if (Number.isNaN(i)) {
                    return 0;
                }
                return i;
            }), 3), line = _b[0], column = _b[1], code = _b[2];
            return {
                line: line,
                column: column,
                expected: expected,
                code: code,
                regexp: new RegExp(regexp),
                matched: false,
            };
        });
        var unmatchedDiags = diagnostics.filter(function (diag) {
            var _a, e_1, _b;
            var line = -1;
            var character = -1;
            if (diag.file && diag.start) {
                (_a = ts.getLineAndCharacterOfPosition(diag.file, diag.start), line = _a.line, character = _a.character);
            }
            var matched = false;
            var msg = formatFn(bazelOpts.target, [diag]);
            // checkDiagMatchesExpected checks if the expected diagnostics matches the
            // actual diagnostics.
            var checkDiagMatchesExpected = function (expDiag, diag) {
                if (expDiag.code !== diag.code || msg.search(expDiag.regexp) === -1) {
                    return false;
                }
                // line and column are optional fields, only check them if they
                // are explicitly specified.
                // line and character are zero based.
                if (expDiag.line !== 0 && expDiag.line !== line + 1) {
                    return false;
                }
                if (expDiag.column !== 0 && expDiag.column !== character + 1) {
                    return false;
                }
                return true;
            };
            try {
                for (var expectedDiags_1 = __values(expectedDiags), expectedDiags_1_1 = expectedDiags_1.next(); !expectedDiags_1_1.done; expectedDiags_1_1 = expectedDiags_1.next()) {
                    var expDiag = expectedDiags_1_1.value;
                    if (checkDiagMatchesExpected(expDiag, diag)) {
                        expDiag.matched = true;
                        matched = true;
                        // continue, one diagnostic may match multiple expected errors.
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (expectedDiags_1_1 && !expectedDiags_1_1.done && (_b = expectedDiags_1.return)) _b.call(expectedDiags_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return !matched;
        });
        var unmatchedErrors = expectedDiags.filter(function (err) { return !err.matched; }).map(function (err) {
            var file = ts.createSourceFile(bazelOpts.target, '/* fake source as marker */', ts.ScriptTarget.Latest);
            var messageText = "Expected a compilation error matching " + JSON.stringify(err.expected);
            return {
                file: file,
                start: 0,
                length: 0,
                messageText: messageText,
                category: ts.DiagnosticCategory.Error,
                code: err.code,
            };
        });
        return unmatchedDiags.concat(unmatchedErrors);
    }
    exports.filterExpected = filterExpected;
    /**
     * Formats the given diagnostics, without pretty printing.  Without colors, it's
     * better for matching against programmatically.
     * @param target The bazel target, e.g. //my/package:target
     */
    function uglyFormat(target, diagnostics) {
        var diagnosticsHost = {
            getCurrentDirectory: function () { return ts.sys.getCurrentDirectory(); },
            getNewLine: function () { return ts.sys.newLine; },
            // Print filenames including their relativeRoot, so they can be located on
            // disk
            getCanonicalFileName: function (f) { return f; }
        };
        return ts.formatDiagnostics(diagnostics, diagnosticsHost);
    }
    exports.uglyFormat = uglyFormat;
    /**
     * Pretty formats the given diagnostics (matching the --pretty tsc flag).
     * @param target The bazel target, e.g. //my/package:target
     */
    function format(target, diagnostics) {
        var diagnosticsHost = {
            getCurrentDirectory: function () { return ts.sys.getCurrentDirectory(); },
            getNewLine: function () { return ts.sys.newLine; },
            // Print filenames including their relativeRoot, so they can be located on
            // disk
            getCanonicalFileName: function (f) { return f; }
        };
        return ts.formatDiagnosticsWithColorAndContext(diagnostics, diagnosticsHost);
    }
    exports.format = format;
});
