// @bazel/typescript/tsc_wrapped/strict_deps.js
/**
 * @license
 * Copyright 2017 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "path", "typescript", "./perf_trace", "./plugin_api"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var path = require("path");
    var ts = require("typescript");
    var perfTrace = require("./perf_trace");
    var pluginApi = require("./plugin_api");
    /** The TypeScript diagnostic code for "Cannot find module ...". */
    exports.TS_ERR_CANNOT_FIND_MODULE = 2307;
    /**
     * The strict_deps plugin checks the imports of the compiled modules.
     *
     * It implements strict deps, i.e. enforces that each file in
     * `config.compilationTargetSrc` only imports from files in
     * `config.allowedStrictDeps`.
     *
     * This is used to implement strict dependency checking -
     * source files in a build target may only import sources of their immediate
     * dependencies, but not sources of their transitive dependencies.
     *
     * strict_deps also makes sure that no imports ends in '.ts'. TypeScript
     * allows imports including the file extension, but our runtime loading support
     * fails with it.
     *
     * strict_deps currently does not check ambient/global definitions.
     */
    exports.PLUGIN = {
        wrap: function (program, config) {
            var proxy = pluginApi.createProxy(program);
            proxy.getSemanticDiagnostics = function (sourceFile) {
                var result = __spread(program.getSemanticDiagnostics(sourceFile));
                perfTrace.wrap('checkModuleDeps', function () {
                    result.push.apply(result, __spread(checkModuleDeps(sourceFile, program.getTypeChecker(), config.allowedStrictDeps, config.rootDir, config.ignoredFilesPrefixes)));
                });
                return result;
            };
            return proxy;
        }
    };
    // Exported for testing
    function checkModuleDeps(sf, tc, allowedDeps, rootDir, ignoredFilesPrefixes) {
        if (ignoredFilesPrefixes === void 0) { ignoredFilesPrefixes = []; }
        var e_1, _a, e_2, _b;
        function stripExt(fn) {
            return fn.replace(/(\.d)?\.tsx?$/, '');
        }
        var allowedMap = {};
        try {
            for (var allowedDeps_1 = __values(allowedDeps), allowedDeps_1_1 = allowedDeps_1.next(); !allowedDeps_1_1.done; allowedDeps_1_1 = allowedDeps_1.next()) {
                var d = allowedDeps_1_1.value;
                allowedMap[stripExt(d)] = true;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (allowedDeps_1_1 && !allowedDeps_1_1.done && (_a = allowedDeps_1.return)) _a.call(allowedDeps_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var result = [];
        var _loop_1 = function (stmt) {
            if (stmt.kind !== ts.SyntaxKind.ImportDeclaration &&
                stmt.kind !== ts.SyntaxKind.ExportDeclaration) {
                return "continue";
            }
            var id = stmt;
            var modSpec = id.moduleSpecifier;
            if (!modSpec)
                return "continue"; // E.g. a bare "export {x};"
            var sym = tc.getSymbolAtLocation(modSpec);
            if (!sym || !sym.declarations || sym.declarations.length < 1) {
                return "continue";
            }
            // Module imports can only have one declaration location.
            var declFileName = sym.declarations[0].getSourceFile().fileName;
            if (allowedMap[stripExt(declFileName)])
                return "continue";
            if (ignoredFilesPrefixes.some(function (p) { return declFileName.startsWith(p); }))
                return "continue";
            var importName = path.posix.relative(rootDir, declFileName);
            result.push({
                file: sf,
                start: modSpec.getStart(),
                length: modSpec.getEnd() - modSpec.getStart(),
                messageText: "transitive dependency on " + importName + " not allowed. " +
                    "Please add the BUILD target to your rule's deps.",
                category: ts.DiagnosticCategory.Error,
                // semantics are close enough, needs taze.
                code: exports.TS_ERR_CANNOT_FIND_MODULE,
            });
        };
        try {
            for (var _c = __values(sf.statements), _d = _c.next(); !_d.done; _d = _c.next()) {
                var stmt = _d.value;
                _loop_1(stmt);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return result;
    }
    exports.checkModuleDeps = checkModuleDeps;
});
