// @bazel/typescript/tsc_wrapped/tsc_wrapped.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "fs", "path", "typescript", "../tsetse/runner", "./cache", "./compiler_host", "./diagnostics", "./manifest", "./perf_trace", "./strict_deps", "./tsconfig", "./worker"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var fs = require("fs");
    var path = require("path");
    var ts = require("typescript");
    var runner_1 = require("../tsetse/runner");
    var cache_1 = require("./cache");
    var compiler_host_1 = require("./compiler_host");
    var bazelDiagnostics = require("./diagnostics");
    var manifest_1 = require("./manifest");
    var perfTrace = require("./perf_trace");
    var strict_deps_1 = require("./strict_deps");
    var tsconfig_1 = require("./tsconfig");
    var worker_1 = require("./worker");
    /**
     * Top-level entry point for tsc_wrapped.
     */
    function main(args) {
        if (worker_1.runAsWorker(args)) {
            worker_1.log('Starting TypeScript compiler persistent worker...');
            worker_1.runWorkerLoop(runOneBuild);
            // Note: intentionally don't process.exit() here, because runWorkerLoop
            // is waiting for async callbacks from node.
        }
        else {
            worker_1.debug('Running a single build...');
            if (args.length === 0)
                throw new Error('Not enough arguments');
            if (!runOneBuild(args)) {
                return 1;
            }
        }
        return 0;
    }
    exports.main = main;
    /** The one ProgramAndFileCache instance used in this process. */
    var cache = new cache_1.ProgramAndFileCache(worker_1.debug);
    function isCompilationTarget(bazelOpts, sf) {
        return (bazelOpts.compilationTargetSrc.indexOf(sf.fileName) !== -1);
    }
    /**
     * Gather diagnostics from TypeScript's type-checker as well as other plugins we
     * install such as strict dependency checking.
     */
    function gatherDiagnostics(options, bazelOpts, program, disabledTsetseRules) {
        // Install extra diagnostic plugins
        if (!bazelOpts.disableStrictDeps) {
            var ignoredFilesPrefixes = [];
            if (bazelOpts.nodeModulesPrefix) {
                // Under Bazel, we exempt external files fetched from npm from strict
                // deps. This is because we allow users to implicitly depend on all the
                // node_modules.
                // TODO(alexeagle): if users opt-in to fine-grained npm dependencies, we
                // should be able to enforce strict deps for them.
                ignoredFilesPrefixes.push(bazelOpts.nodeModulesPrefix);
                if (options.rootDir) {
                    ignoredFilesPrefixes.push(path.resolve(options.rootDir, 'node_modules'));
                }
            }
            program = strict_deps_1.PLUGIN.wrap(program, __assign({}, bazelOpts, { rootDir: options.rootDir, ignoredFilesPrefixes: ignoredFilesPrefixes }));
        }
        if (!bazelOpts.isJsTranspilation) {
            var selectedTsetsePlugin = runner_1.PLUGIN;
            program = selectedTsetsePlugin.wrap(program, disabledTsetseRules);
        }
        // TODO(alexeagle): support plugins registered by config
        var diagnostics = [];
        perfTrace.wrap('type checking', function () {
            var e_1, _a;
            // These checks mirror ts.getPreEmitDiagnostics, with the important
            // exception of avoiding b/30708240, which is that if you call
            // program.getDeclarationDiagnostics() it somehow corrupts the emit.
            perfTrace.wrap("global diagnostics", function () {
                diagnostics.push.apply(diagnostics, __spread(program.getOptionsDiagnostics()));
                diagnostics.push.apply(diagnostics, __spread(program.getGlobalDiagnostics()));
            });
            var sourceFilesToCheck;
            if (bazelOpts.typeCheckDependencies) {
                sourceFilesToCheck = program.getSourceFiles();
            }
            else {
                sourceFilesToCheck = program.getSourceFiles().filter(function (f) { return isCompilationTarget(bazelOpts, f); });
            }
            var _loop_1 = function (sf) {
                perfTrace.wrap("check " + sf.fileName, function () {
                    diagnostics.push.apply(diagnostics, __spread(program.getSyntacticDiagnostics(sf)));
                    diagnostics.push.apply(diagnostics, __spread(program.getSemanticDiagnostics(sf)));
                });
                perfTrace.snapshotMemoryUsage();
            };
            try {
                for (var sourceFilesToCheck_1 = __values(sourceFilesToCheck), sourceFilesToCheck_1_1 = sourceFilesToCheck_1.next(); !sourceFilesToCheck_1_1.done; sourceFilesToCheck_1_1 = sourceFilesToCheck_1.next()) {
                    var sf = sourceFilesToCheck_1_1.value;
                    _loop_1(sf);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (sourceFilesToCheck_1_1 && !sourceFilesToCheck_1_1.done && (_a = sourceFilesToCheck_1.return)) _a.call(sourceFilesToCheck_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
        return diagnostics;
    }
    exports.gatherDiagnostics = gatherDiagnostics;
    /**
     * Runs a single build, returning false on failure.  This is potentially called
     * multiple times (once per bazel request) when running as a bazel worker.
     * Any encountered errors are written to stderr.
     */
    function runOneBuild(args, inputs) {
        var e_2, _a;
        if (args.length !== 1) {
            console.error('Expected one argument: path to tsconfig.json');
            return false;
        }
        perfTrace.snapshotMemoryUsage();
        // Strip leading at-signs, used in build_defs.bzl to indicate a params file
        var tsconfigFile = args[0].replace(/^@+/, '');
        var _b = __read(tsconfig_1.parseTsconfig(tsconfigFile), 3), parsed = _b[0], errors = _b[1], target = _b[2].target;
        if (errors) {
            console.error(bazelDiagnostics.format(target, errors));
            return false;
        }
        if (!parsed) {
            throw new Error('Impossible state: if parseTsconfig returns no errors, then parsed should be non-null');
        }
        var options = parsed.options, bazelOpts = parsed.bazelOpts, files = parsed.files, disabledTsetseRules = parsed.disabledTsetseRules;
        if (bazelOpts.maxCacheSizeMb !== undefined) {
            var maxCacheSizeBytes = bazelOpts.maxCacheSizeMb * (1 << 20);
            cache.setMaxCacheSize(maxCacheSizeBytes);
        }
        else {
            cache.resetMaxCacheSize();
        }
        var fileLoader;
        if (inputs) {
            fileLoader = new cache_1.CachedFileLoader(cache);
            // Resolve the inputs to absolute paths to match TypeScript internals
            var resolvedInputs = new Map();
            try {
                for (var _c = __values(Object.keys(inputs)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var key = _d.value;
                    resolvedInputs.set(tsconfig_1.resolveNormalizedPath(key), inputs[key]);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_2) throw e_2.error; }
            }
            cache.updateCache(resolvedInputs);
        }
        else {
            fileLoader = new cache_1.UncachedFileLoader();
        }
        var perfTracePath = bazelOpts.perfTracePath;
        if (!perfTracePath) {
            return runFromOptions(fileLoader, options, bazelOpts, files, disabledTsetseRules);
        }
        worker_1.log('Writing trace to', perfTracePath);
        var success = perfTrace.wrap('runOneBuild', function () { return runFromOptions(fileLoader, options, bazelOpts, files, disabledTsetseRules); });
        if (!success)
            return false;
        // Force a garbage collection pass.  This keeps our memory usage
        // consistent across multiple compilations, and allows the file
        // cache to use the current memory usage as a guideline for expiring
        // data.  Note: this is intentionally not within runFromOptions(), as
        // we want to gc only after all its locals have gone out of scope.
        global.gc();
        perfTrace.snapshotMemoryUsage();
        perfTrace.write(perfTracePath);
        return true;
    }
    // We only allow our own code to use the expected_diagnostics attribute
    var expectDiagnosticsWhitelist = [];
    function runFromOptions(fileLoader, options, bazelOpts, files, disabledTsetseRules) {
        perfTrace.snapshotMemoryUsage();
        cache.resetStats();
        cache.traceStats();
        var compilerHostDelegate = ts.createCompilerHost({ target: ts.ScriptTarget.ES5 });
        var moduleResolver = bazelOpts.isJsTranspilation ?
            makeJsModuleResolver(bazelOpts.workspaceName) :
            ts.resolveModuleName;
        var compilerHost = new compiler_host_1.CompilerHost(files, options, bazelOpts, compilerHostDelegate, fileLoader, moduleResolver);
        var oldProgram = cache.getProgram(bazelOpts.target);
        var program = perfTrace.wrap('createProgram', function () { return ts.createProgram(compilerHost.inputFiles, options, compilerHost, oldProgram); });
        cache.putProgram(bazelOpts.target, program);
        if (!bazelOpts.isJsTranspilation) {
            // If there are any TypeScript type errors abort now, so the error
            // messages refer to the original source.  After any subsequent passes
            // (decorator downleveling or tsickle) we do not type check.
            var diagnostics_1 = gatherDiagnostics(options, bazelOpts, program, disabledTsetseRules);
            if (!expectDiagnosticsWhitelist.length ||
                expectDiagnosticsWhitelist.some(function (p) { return bazelOpts.target.startsWith(p); })) {
                diagnostics_1 = bazelDiagnostics.filterExpected(bazelOpts, diagnostics_1, bazelDiagnostics.uglyFormat);
            }
            else if (bazelOpts.expectedDiagnostics.length > 0) {
                console.error("Only targets under " + expectDiagnosticsWhitelist.join(', ') + " can use " +
                    'expected_diagnostics, but got', bazelOpts.target);
            }
            if (diagnostics_1.length > 0) {
                console.error(bazelDiagnostics.format(bazelOpts.target, diagnostics_1));
                worker_1.debug('compilation failed at', new Error().stack);
                return false;
            }
        }
        var compilationTargets = program.getSourceFiles().filter(function (fileName) { return isCompilationTarget(bazelOpts, fileName); });
        var diagnostics = [];
        var useTsickleEmit = bazelOpts.tsickle;
        if (useTsickleEmit) {
            diagnostics = emitWithTsickle(program, compilerHost, compilationTargets, options, bazelOpts);
        }
        else {
            diagnostics = emitWithTypescript(program, compilationTargets);
        }
        if (diagnostics.length > 0) {
            console.error(bazelDiagnostics.format(bazelOpts.target, diagnostics));
            worker_1.debug('compilation failed at', new Error().stack);
            return false;
        }
        cache.printStats();
        return true;
    }
    function emitWithTypescript(program, compilationTargets) {
        var e_3, _a;
        var diagnostics = [];
        try {
            for (var compilationTargets_1 = __values(compilationTargets), compilationTargets_1_1 = compilationTargets_1.next(); !compilationTargets_1_1.done; compilationTargets_1_1 = compilationTargets_1.next()) {
                var sf = compilationTargets_1_1.value;
                var result = program.emit(sf);
                diagnostics.push.apply(diagnostics, __spread(result.diagnostics));
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (compilationTargets_1_1 && !compilationTargets_1_1.done && (_a = compilationTargets_1.return)) _a.call(compilationTargets_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return diagnostics;
    }
    function emitWithTsickle(program, compilerHost, compilationTargets, options, bazelOpts) {
        var e_4, _a;
        var emitResults = [];
        var diagnostics = [];
        // The 'tsickle' import above is only used in type positions, so it won't
        // result in a runtime dependency on tsickle.
        // If the user requests the tsickle emit, then we dynamically require it
        // here for use at runtime.
        var optTsickle;
        try {
            // tslint:disable-next-line:no-require-imports
            optTsickle = require('tsickle');
        }
        catch (e) {
            if (e.code !== 'MODULE_NOT_FOUND') {
                throw e;
            }
            throw new Error('When setting bazelOpts { tsickle: true }, ' +
                'you must also add a devDependency on the tsickle npm package');
        }
        perfTrace.wrap('emit', function () {
            var e_5, _a;
            var _loop_2 = function (sf) {
                perfTrace.wrap("emit " + sf.fileName, function () {
                    emitResults.push(optTsickle.emitWithTsickle(program, compilerHost, compilerHost, options, sf));
                });
            };
            try {
                for (var compilationTargets_3 = __values(compilationTargets), compilationTargets_3_1 = compilationTargets_3.next(); !compilationTargets_3_1.done; compilationTargets_3_1 = compilationTargets_3.next()) {
                    var sf = compilationTargets_3_1.value;
                    _loop_2(sf);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (compilationTargets_3_1 && !compilationTargets_3_1.done && (_a = compilationTargets_3.return)) _a.call(compilationTargets_3);
                }
                finally { if (e_5) throw e_5.error; }
            }
        });
        var emitResult = optTsickle.mergeEmitResults(emitResults);
        diagnostics.push.apply(diagnostics, __spread(emitResult.diagnostics));
        // If tsickle reported diagnostics, don't produce externs or manifest outputs.
        if (diagnostics.length > 0) {
            return diagnostics;
        }
        var externs = '/** @externs */\n' +
            '// generating externs was disabled using generate_externs=False\n';
        if (bazelOpts.tsickleGenerateExterns) {
            externs =
                optTsickle.getGeneratedExterns(emitResult.externs, options.rootDir);
        }
        if (bazelOpts.tsickleExternsPath) {
            // Note: when tsickleExternsPath is provided, we always write a file as a
            // marker that compilation succeeded, even if it's empty (just containing an
            // @externs).
            fs.writeFileSync(bazelOpts.tsickleExternsPath, externs);
            // When generating externs, generate an externs file for each of the input
            // .d.ts files.
            if (bazelOpts.tsickleGenerateExterns &&
                compilerHost.provideExternalModuleDtsNamespace) {
                try {
                    for (var compilationTargets_2 = __values(compilationTargets), compilationTargets_2_1 = compilationTargets_2.next(); !compilationTargets_2_1.done; compilationTargets_2_1 = compilationTargets_2.next()) {
                        var extern = compilationTargets_2_1.value;
                        if (!extern.isDeclarationFile)
                            continue;
                        var outputBaseDir = options.outDir;
                        var relativeOutputPath = compilerHost.relativeOutputPath(extern.fileName);
                        mkdirp(outputBaseDir, path.dirname(relativeOutputPath));
                        var outputPath = path.join(outputBaseDir, relativeOutputPath);
                        var moduleName = compilerHost.pathToModuleName('', extern.fileName);
                        fs.writeFileSync(outputPath, "goog.module('" + moduleName + "');\n" +
                            "// Export an empty object of unknown type to allow imports.\n" +
                            "// TODO: use typeof once available\n" +
                            "exports = /** @type {?} */ ({});\n");
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (compilationTargets_2_1 && !compilationTargets_2_1.done && (_a = compilationTargets_2.return)) _a.call(compilationTargets_2);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
            }
        }
        if (bazelOpts.manifest) {
            perfTrace.wrap('manifest', function () {
                var manifest = manifest_1.constructManifest(emitResult.modulesManifest, compilerHost);
                fs.writeFileSync(bazelOpts.manifest, manifest);
            });
        }
        return diagnostics;
    }
    /**
     * Creates directories subdir (a slash separated relative path) starting from
     * base.
     */
    function mkdirp(base, subdir) {
        var steps = subdir.split(path.sep);
        var current = base;
        for (var i = 0; i < steps.length; i++) {
            current = path.join(current, steps[i]);
            if (!fs.existsSync(current))
                fs.mkdirSync(current);
        }
    }
    /**
     * Resolve module filenames for JS modules.
     *
     * JS module resolution needs to be different because when transpiling JS we
     * do not pass in any dependencies, so the TS module resolver will not resolve
     * any files.
     *
     * Fortunately, JS module resolution is very simple. The imported module name
     * must either a relative path, or the workspace root (i.e. 'google3'),
     * so we can perform module resolution entirely based on file names, without
     * looking at the filesystem.
     */
    function makeJsModuleResolver(workspaceName) {
        // The literal '/' here is cross-platform safe because it's matching on
        // import specifiers, not file names.
        var workspaceModuleSpecifierPrefix = workspaceName + "/";
        var workspaceDir = "" + path.sep + workspaceName + path.sep;
        function jsModuleResolver(moduleName, containingFile, compilerOptions, host) {
            var resolvedFileName;
            if (containingFile === '') {
                // In tsickle we resolve the filename against '' to get the goog module
                // name of a sourcefile.
                resolvedFileName = moduleName;
            }
            else if (moduleName.startsWith(workspaceModuleSpecifierPrefix)) {
                // Given a workspace name of 'foo', we want to resolve import specifiers
                // like: 'foo/project/file.js' to the absolute filesystem path of
                // project/file.js within the workspace.
                var workspaceDirLocation = containingFile.indexOf(workspaceDir);
                if (workspaceDirLocation < 0) {
                    return { resolvedModule: undefined };
                }
                var absolutePathToWorkspaceDir = containingFile.slice(0, workspaceDirLocation);
                resolvedFileName = path.join(absolutePathToWorkspaceDir, moduleName);
            }
            else {
                if (!moduleName.startsWith('./') && !moduleName.startsWith('../')) {
                    throw new Error("Unsupported module import specifier: " + JSON.stringify(moduleName) + ".\n" +
                        "JS module imports must either be relative paths " +
                        "(beginning with '.' or '..'), " +
                        ("or they must begin with '" + workspaceName + "/'."));
                }
                resolvedFileName = path.join(path.dirname(containingFile), moduleName);
            }
            return {
                resolvedModule: {
                    resolvedFileName: resolvedFileName,
                    extension: ts.Extension.Js,
                    // These two fields are cargo culted from what ts.resolveModuleName
                    // seems to return.
                    packageId: undefined,
                    isExternalLibraryImport: false,
                }
            };
        }
        return jsModuleResolver;
    }
    if (require.main === module) {
        // Do not call process.exit(), as that terminates the binary before
        // completing pending operations, such as writing to stdout or emitting the
        // v8 performance log. Rather, set the exit code and fall off the main
        // thread, which will cause node to terminate cleanly.
        process.exitCode = main(process.argv.slice(2));
    }
});
