// @bazel/typescript/tsc_wrapped/worker.js
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "path"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var path = require("path");
    /* tslint:disable:no-require-imports */
    var protobufjs = require('protobufjs');
    // tslint:disable-next-line:variable-name: ByteBuffer is instantiatable.
    var ByteBuffer = require('bytebuffer');
    protobufjs.convertFieldsToCamelCase = true;
    exports.DEBUG = false;
    function debug() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (exports.DEBUG)
            console.error.apply(console, args);
    }
    exports.debug = debug;
    /**
     * Write a message to stderr, which appears in the bazel log and is visible to
     * the end user.
     */
    function log() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.error.apply(console, args);
    }
    exports.log = log;
    function runAsWorker(args) {
        return args.indexOf('--persistent_worker') !== -1;
    }
    exports.runAsWorker = runAsWorker;
    var workerpb = (function loadWorkerPb() {
        var protoPath = '../worker_protocol.proto';
        // Use node module resolution so we can find the .proto file in any of the root dirs
        var protofile;
        try {
            // Look for the .proto file relative in its @bazel/typescript npm package
            // location
            protofile = require.resolve(protoPath);
        }
        catch (e) {
        }
        if (!protofile) {
            // If not found above, look for the .proto file in its rules_typescript
            // workspace location
            // This extra lookup should never happen in google3. It's only needed for
            // local development in the rules_typescript repo.
            protofile = require.resolve('../../third_party/github.com/bazelbuild/bazel/src/main/protobuf/worker_protocol.proto');
        }
        // Under Bazel, we use the version of TypeScript installed in the user's
        // workspace This means we also use their version of protobuf.js. Handle both.
        // v5 and v6 by checking which one is present.
        if (protobufjs.loadProtoFile) {
            // Protobuf.js v5
            var protoNamespace = protobufjs.loadProtoFile(protofile);
            if (!protoNamespace) {
                throw new Error('Cannot find ' + path.resolve(protoPath));
            }
            return protoNamespace.build('blaze.worker');
        }
        else {
            // Protobuf.js v6
            var protoNamespace = protobufjs.loadSync(protofile);
            if (!protoNamespace) {
                throw new Error('Cannot find ' + path.resolve(protoPath));
            }
            return protoNamespace.lookup('blaze.worker');
        }
    })();
    function runWorkerLoop(runOneBuild) {
        // Hook all output to stderr and write it to a buffer, then include
        // that buffer's in the worker protcol proto's textual output.  This
        // means you can log via console.error() and it will appear to the
        // user as expected.
        var consoleOutput = '';
        process.stderr.write =
            function (chunk) {
                var otherArgs = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    otherArgs[_i - 1] = arguments[_i];
                }
                consoleOutput += chunk.toString();
                return true;
            };
        // Accumulator for asynchronously read input.
        // tslint:disable-next-line:no-any protobufjs is untyped
        var buf;
        process.stdin.on('readable', function () {
            var e_1, _a;
            var chunk = process.stdin.read();
            if (!chunk)
                return;
            var wrapped = ByteBuffer.wrap(chunk);
            buf = buf ? ByteBuffer.concat([buf, wrapped]) : wrapped;
            try {
                var req = void 0;
                // Read all requests that have accumulated in the buffer.
                while ((req = workerpb.WorkRequest.decodeDelimited(buf)) != null) {
                    debug('=== Handling new build request');
                    // Reset accumulated log output.
                    consoleOutput = '';
                    var args = req.getArguments();
                    var inputs = {};
                    try {
                        for (var _b = __values(req.getInputs()), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var input = _c.value;
                            inputs[input.getPath()] = input.getDigest().toString('hex');
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    debug('Compiling with:\n\t' + args.join('\n\t'));
                    var exitCode = runOneBuild(args, inputs) ? 0 : 1;
                    process.stdout.write(new workerpb.WorkResponse()
                        .setExitCode(exitCode)
                        .setOutput(consoleOutput)
                        .encodeDelimited()
                        .toBuffer());
                    // Force a garbage collection pass.  This keeps our memory usage
                    // consistent across multiple compilations, and allows the file
                    // cache to use the current memory usage as a guideline for expiring
                    // data.  Note: this is intentionally not within runOneBuild(), as
                    // we want to gc only after all its locals have gone out of scope.
                    global.gc();
                }
                // Avoid growing the buffer indefinitely.
                buf.compact();
            }
            catch (e) {
                log('Compilation failed', e.stack);
                process.stdout.write(new workerpb.WorkResponse()
                    .setExitCode(1)
                    .setOutput(consoleOutput)
                    .encodeDelimited()
                    .toBuffer());
                // Clear buffer so the next build won't read an incomplete request.
                buf = null;
            }
        });
    }
    exports.runWorkerLoop = runWorkerLoop;
});
