// @bazel/karma/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@bazel/karma", ["require", "exports", "crypto", "fs", "path", "process", "tmp"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /*
     * Concat all JS files before serving.
     */
    var crypto = require("crypto");
    var fs = require("fs");
    var path = require("path");
    var process = require("process");
    var tmp = require("tmp");
    ///<reference types="lib.dom"/>
    /**
     * Return SHA1 of data buffer.
     */
    function sha1(data) {
        var hash = crypto.createHash('sha1');
        hash.update(data);
        return hash.digest('hex');
    }
    /**
     * Entry-point for the Karma plugin.
     */
    function initConcatJs(logger, emitter, basePath) {
        var log = logger.create('framework.concat_js');
        // Create a tmp file for the concat bundle that is automatically cleaned up on
        // exit.
        var tmpFile = tmp.fileSync({ keep: false, dir: process.env['TEST_TMPDIR'] });
        emitter.on('file_list_modified', function (files) {
            var bundleFile = {
                path: '/concatjs_bundle.js',
                contentPath: tmpFile.name,
                isUrl: false,
                content: ''
            };
            var included = [];
            files.included.forEach(function (file) {
                if (path.extname(file.originalPath) !== '.js') {
                    // Preserve all non-JS that were there in the included list.
                    included.push(file);
                }
                else {
                    var relativePath = path.relative(basePath, file.originalPath).replace(/\\/g, '/');
                    // Remove 'use strict'.
                    var content = file.content.replace(/('use strict'|"use strict");?/, '');
                    content = JSON.stringify(content + '\n//# sourceURL=http://concatjs/base/' +
                        relativePath + '\n');
                    content = "//" + relativePath + "\neval(" + content + ");\n";
                    bundleFile.content += content;
                }
            });
            bundleFile.sha = sha1(Buffer.from(bundleFile.content));
            bundleFile.mtime = new Date();
            included.unshift(bundleFile);
            files.included = included;
            files.served.push(bundleFile);
            log.debug('Writing concatjs bundle to tmp file %s', bundleFile.contentPath);
            fs.writeFileSync(bundleFile.contentPath, bundleFile.content);
        });
    }
    initConcatJs.$inject = ['logger', 'emitter', 'config.basePath'];
    module.exports = {
        'framework:concat_js': ['factory', initConcatJs]
    };
});
